File: pset2/crack/crack.c
#define _XOPEN_SOURCE
#include <unistd.h>
#include <cs50.h>
#include <stdio.h>
#include <crypt.h>
#include <string.h>

// trying all possibile password
bool tryPassword(char* plain, char* cipher, char* salt)
{   
    char dict[] = {"0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"};
    int dictLength = strlen(dict);
    
    // variable for representing each digit
    int first = 0;
    int second = 1;
    int third = 2;
    int forth = 3;
    
    // ensure a proper end of each string
    dict[4] = '\0';
    salt[2] = '\0';
    cipher[4] = '\0';
    
    // trying password XXXX
    for(int i = 0 ; i < dictLength ; i++ )
    {
        cipher[first]=dict[i];
        for(int j = 0 ; j < dictLength ; j++)
        {
            cipher[second]=dict[j];
            for(int k = 0 ; k < dictLength ; k++)
            {
                cipher[third]=dict[k];
                for(int l = 0 ; l < dictLength ; l++)
                {
                    cipher[forth]=dict[l];
                    if( ! strcmp ( crypt(cipher, salt), plain ) ) 
                    {
                        return true;
                    }
                }
                
            }
        }
    }
    
    // trying password XXX
    cipher[forth]='\0';
    for(int i = 0 ; i < dictLength ; i++ )
    {
        cipher[first]=dict[i];
        for(int j = 0 ; j < dictLength ; j++)
        {
            cipher[second]=dict[j];
            for(int k = 0 ; k < dictLength ; k++)
            {
                cipher[third]=dict[k];
                if( ! strcmp ( crypt(cipher, salt), plain ) ) 
                {
                    return true;
                }
            }
        }
    }
    // trying password XX
    cipher[third]='\0';
    for(int i = 0 ; i < dictLength ; i++ )
    {
        cipher[first]=dict[i];
        for(int j = 0 ; j < dictLength ; j++)
        {
            cipher[second]=dict[j];
            if( ! strcmp ( crypt(cipher, salt), plain ) ) 
            {
                return true;
            }   
        }
    }
    
    // trying password X
    cipher[second]='\0';
    for(int i = 0 ; i < dictLength ; i++)
    {
        cipher[first]=dict[i];
        if( ! strcmp ( crypt(cipher, salt), plain ) ) 
        {
            return true;
        }   
    }
    
    return false;
}

int main(int argc, string argv[])
{
    // ensure proper usage
    if ( argc < 2 || argc > 2)
    {
        printf("Illegal input\n");
        exit(1);
    }
    
    char salt[2];
    
    // get salt from input
    strncpy(salt, argv[1], 2);
    
    int plainLength = strlen(argv[1]);
    char plaintext[plainLength];
    // get plain text from input
    strcpy(plaintext,argv[1]);
    char ciphertext[4];
    
    // trying
    if(tryPassword(plaintext,ciphertext,salt))
    {
        // success
        //printf("Matched\n");
        printf("%s\n",ciphertext);
        return 0;
    }
    
    // failure
    //printf("Not Match\n");
    
    return 1;
}

File: pset6/crack.py
import crypt
import sys
import cs50

# function to test all password
def tryPassword(ciphertext, salt):
    # dictionary for all possible letters
    dictionary = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    # Compare
    plaintext = ""
    # save each word
    password = ""
    # _ test case only one letter 
    for letter in dictionary:
        password = letter
        plaintext = crypt.crypt(password, salt)
        if plaintext == ciphertext:
            return password
    # __ test case two letters
    password = ""
    plaintext = ""
    for letter1 in dictionary:
        password += letter1
        for letter2 in dictionary:
            password += letter2
            plaintext = crypt.crypt(password, salt)
            if plaintext == ciphertext:
                return password
            password = password[0:1]
        password = password[:0]
    # ___ test case three letters
    password = ""
    plaintext = ""
    for letter1 in dictionary:
        password += letter1
        for letter2 in dictionary:
            password += letter2
            for letter3 in dictionary:
                password += letter3
                plaintext = crypt.crypt(password, salt)
                if plaintext == ciphertext:
                    return password
                password = password[0:2]
            password = password[0:1]
        password = password[:0]
    # ____ test case four letters
    password = ""
    plaintext = ""
    for letter1 in dictionary:
        password += letter1
        for letter2 in dictionary:
            password += letter2
            for letter3 in dictionary:
                password += letter3
                for letter4 in dictionary:
                    password += letter4
                    plaintext = crypt.crypt(password, salt)
                    if plaintext == ciphertext:
                        return password
                    password = password[0:3]
                password = password[0:2]
            password = password[0:1]
        password = password[:0]
    
    # no match password
    return ""

def main():
    # ensure proper usage
    if len(sys.argv) != 2:
        sys.exit("Usage: ./crack hash")
    # get ciphertext
    ciphertext = sys.argv[1]
    # get salt
    salt = ciphertext[:2]
    # test - > print("ciphertext-> {} salt-> {}".format(ciphertext,salt))
    plaintext = tryPassword(ciphertext, salt)
    print(plaintext)
    return

if __name__ == '__main__':
    main()